core/platform/email/gae_email_services_test.py

```diff
@@ -38,8 +44,166 @@ def test_sending_email(self):
+                feconf.SYSTEM_EMAIL_ADDRESS, feconf.ADMIN_EMAIL_ADDRESS,
+                'subject', 'body', 'html')
+        messages = self.mail_stub.get_sent_messages(
+            to=feconf.ADMIN_EMAIL_ADDRESS)
+        self.assertFalse(hasattr(messages, 'reply_to'))
+
+    def test_that_email_not_sent_if_sender_address_is_malformed(self):
+        malformed_sender_email = ''
+        email_exception = (
+            self.assertRaisesRegexp(
+                ValueError, 'Malformed sender email address: %s'
+                % malformed_sender_email))
+        with self.swap(feconf, 'CAN_SEND_EMAILS', True), email_exception:
+            gae_email_services.send_mail(
+                malformed_sender_email, self.RECIPIENT_EMAIL,
+                'subject', 'body', 'html')
+        messages = self.mail_stub.get_sent_messages(
+            to=self.RECIPIENT_EMAIL)
+        self.assertEqual(0, len(messages))
+
+    def test_that_email_not_sent_if_recipient_address_is_malformed(self):
+        malformed_recipient_email = ''
+        email_exception = (self.assertRaisesRegexp(
+            ValueError, 'Malformed recipient email address: %s'
+            % malformed_recipient_email))
+        with self.swap(feconf, 'CAN_SEND_EMAILS', True), email_exception:
+            gae_email_services.send_mail(
+                feconf.ADMIN_EMAIL_ADDRESS, malformed_recipient_email,
+                'subject', 'body', 'html')
+
+
+class BulkEmailsTests(test_utils.GenericTestBase):
+    SENDER_EMAIL = 'sender@example.com'
+    SENDER_USERNAME = 'sender'
+    RECIPIENT_A_EMAIL = 'a@example.com'
+    RECIPIENT_A_USERNAME = 'usera'
+    RECIPIENT_B_EMAIL = 'b@example.com'
+    RECIPIENT_B_USERNAME = 'userb'
+    RECIPIENT_EMAILS = [RECIPIENT_A_EMAIL, RECIPIENT_B_EMAIL]
+
+    def setUp(self):
+        super(BulkEmailsTests, self).setUp()
+        # SENDER is authorised sender.
+        # A and B are recipients.
+        self.signup(self.SENDER_EMAIL, self.SENDER_USERNAME)
+        self.signup(self.RECIPIENT_A_EMAIL, self.RECIPIENT_A_USERNAME)
+        self.signup(self.RECIPIENT_B_EMAIL, self.RECIPIENT_B_USERNAME)
+        self.recipient_emails = [self.RECIPIENT_A_EMAIL, self.RECIPIENT_B_EMAIL]
+
+    def test_that_correct_bulk_emails_sent(self):
+        with self.swap(feconf, 'CAN_SEND_EMAILS', True):
+            gae_email_services.send_bulk_mail(
+                self.SENDER_EMAIL, self.RECIPIENT_EMAILS,
+                'subject', 'body', 'html')
+        message_a = self.mail_stub.get_sent_messages(
+            to=self.RECIPIENT_EMAILS[0])
+        self.assertEqual(len(message_a), 1)
+        message_b = self.mail_stub.get_sent_messages(
+            to=self.RECIPIENT_EMAILS[1])
+        self.assertEqual(len(message_b), 1)
+
+    def test_that_bulk_emails_not_sent_if_can_send_emails_is_false(self):
+        # Emails are not sent if the CAN_SEND_EMAILS setting is not turned on.
+        email_exception = (self.assertRaisesRegexp(
```

Comment:
Drop outer parens, they are not needed.

You might also consider just inlining this. It reduces the need to keep an additional variable in one's head while reading the code. We do similar things elsewhere so you can look at the existing code for examples. (Ditto below and elsewhere.)

Reply:
Fixed